openid-4-verifiable-presentations-1_0.md

```diff
@@ -432,54 +432,26 @@ Note: "https://self-issued.me/v2" is a symbolic string and can be used as an `au
 
 ## Verifier Metadata Management {#client_metadata_management}
 
-The `client_id_scheme` enables deployments of this specification to use different mechanisms to obtain and validate metadata of the Verifier beyond the scope of [@!RFC6749]. The term `client_id_scheme` is used since the Verifier is acting as an OAuth 2.0 Client.
+This specification defines additional mechanisms to enable Wallets to determine whether the provided `redirect_uri` and `response_uri` parameters and public verification keys belong to a particular Client, utilizing existing mechanisms such as Dynamic Client Registration [@!RFC7591], OpenID Federation [@!OpenID.Federation], and the `client_metadata` parameter defined by this specification. Additional Verifier Metadata is retrieved using existing mechanisms or the `client_metadata` parameter.
 
-This specification defines the following values for the `client_id_scheme` parameter, followed by the examples where applicable: 
+The Wallet MUST verify that the provided `redirect_uri` or `response_uri` belongs to the Client.
 
-* `pre-registered`: This value represents the [@!RFC6749] default behavior, i.e., the Client Identifier needs to be known to the Wallet in advance of the Authorization Request. The Verifier metadata is obtained using [@!RFC7591] or through out-of-band mechanisms.
+If the Authorization Request is signed, the Wallet MUST authenticate the Request Object against the `client_id` contained in the Request Object. In this case, any `jwks` parameter in the `client_metadata` parameter MUST NOT be used to verify the signature of the Request Object.
 
-* `redirect_uri`: This value indicates that the Verifier's Redirect URI (or Response URI when Response Mode `direct_post` is used) is also the value of the Client Identifier. The Authorization Request MUST NOT be signed. The Verifier MAY omit the `redirect_uri` Authorization Request parameter (or `response_uri` when Response Mode `direct_post` is used). All Verifier metadata parameters MUST be passed using the `client_metadata` parameter defined in (#vp_token_request).
+If the Verifier Metadata is not obtained from a trusted source and if the Authorization Request is unsigned, the Wallet MUST verify that the `client_id` matches the `redirect_uri` or `response_uri` parameter. Examples for trusted sources include prior registration (e.g., using [@!RFC7591]) or other out-of-band processes that allow the Wallet to retrieve trusted Verifier Metadata for a particular Client.
 
-The following is a non-normative example of a request when `client_id` equals `redirect_uri`.
+The following is a list of methods defined by this specification that MAY be used by the Wallet to authenticate the Request Object and to verify that the `redirect_uri` or `response_uri` belongs to the Client:
```

Comment:
It states that the following mechanisms MAY be used, but the mechanisms themselves contain MUST. What does this mean exactly, some options:

1. It's optional to use these mechanisms even when the conditions are applicable, but if the mechanism is used by the wallet, the requirements have to be followed.
2. When the conditions are applicable the mechanism MUST be used, otherwise an error must be given.
3. When the conditions are applicable the mechanism MUST be used, otherwise the client is considered not authenticated, with the consequences being out of scope.

Something else?

Reply:
The PR aims to allow for Option 1. I agree that the combination of MAY and MUST is not clear.

Reply:
What happens if the request has contradicting properties, e.g., fulfills the rules in X.509 and Federation at the same time?